homeassistant/components/integration/sensor.py

```diff
@@ -491,21 +492,33 @@ def _integrate_on_state_change(
             self.async_write_ha_state()
             return
 
+        if old_state:
+            # state has changed, we recover old_state from the event
+            old_state_state = old_state.state
+            old_last_reported = old_state.last_reported
+        else:
+            # event state reported without any state change
+            old_state_state = new_state.state
+
         self._attr_available = True
         self._derive_and_set_attributes_from_state(new_state)
 
-        if old_state is None:
+        if old_last_reported is None and old_state is None:
             self.async_write_ha_state()
             return
 
-        if not (states := self._method.validate_states(old_state, new_state)):
+        if not (
+            states := self._method.validate_states(old_state_state, new_state.state)
+        ):
             self.async_write_ha_state()
             return
 
+        if TYPE_CHECKING:
+            assert old_last_reported is not None
```

Comment:
Where is there a guarantee of an old last reported? Can mypy handle an xor assert earlier to assert that just old or last old exists and not the other? Might be a more clear assert.

Reply:
Python's typed dicts kind of suck. If you have an idea for a clearer assert, I'm happy to change it. Or maybe suggest a comment which makes it clearer?